File: app/src/main/kotlin/com/agentdeck/data/repository/TaskRepository.kt
// SPDX-License-Identifier: MIT
package com.agentdeck.data.repository

import com.agentdeck.core.domain.*
import com.agentdeck.data.database.TaskDao
import com.agentdeck.data.database.TaskEntity
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class TaskRepository(
    private val dao: TaskDao
) {
    private val json = Json { ignoreUnknownKeys = true }
    
    fun getForBlueprint(blueprintId: String): Flow<List<Task>> {
        return dao.getForBlueprint(blueprintId).map { entities ->
            entities.map { it.toDomain() }
        }
    }
    
    fun getByStatus(status: TaskStatus): Flow<List<Task>> {
        return dao.getByStatus(status.name).map { entities ->
            entities.map { it.toDomain() }
        }
    }
    
    suspend fun getById(id: String): Task? {
        return dao.getById(id)?.toDomain()
    }
    
    suspend fun saveAll(tasks: List<Task>) {
        dao.insertAll(tasks.map { it.toEntity() })
    }

    @Deprecated("blueprintId ist jetzt Feld in Task - direkt in Task setzen", ReplaceWith("saveAll(tasks.map { it.copy(blueprintId = blueprintId) })"))
    suspend fun saveAll(tasks: List<Task>, blueprintId: String) {
        dao.insertAll(tasks.map { 
            val t = if (it.blueprintId == "default" && blueprintId != "default") it.copy(blueprintId = blueprintId) else it
            t.toEntity() 
        })
    }

    suspend fun saveAllWithDefault(tasks: List<Task>) = saveAll(tasks)
    
    suspend fun update(task: Task) {
        dao.update(task.toEntity())
    }

    @Deprecated("blueprintId ist jetzt Feld in Task", ReplaceWith("update(task.copy(blueprintId = blueprintId))"))
    suspend fun update(task: Task, blueprintId: String) {
        val t = if (task.blueprintId == "default" && blueprintId != "default") task.copy(blueprintId = blueprintId) else task
        dao.update(t.toEntity())
    }
    
    suspend fun updateStatus(taskId: String, status: TaskStatus) {
        dao.updateStatus(taskId, status.name)
    }
    
    suspend fun delete(task: Task) {
        dao.delete(task.toEntity())
    }

    @Deprecated("blueprintId ist jetzt Feld in Task", ReplaceWith("delete(task.copy(blueprintId = blueprintId))"))
    suspend fun delete(task: Task, blueprintId: String) {
        val t = if (task.blueprintId == "default" && blueprintId != "default") task.copy(blueprintId = blueprintId) else task
        dao.delete(t.toEntity())
    }
    
    private fun Task.toEntity(): TaskEntity {
        return TaskEntity(
            id = id,
            blueprintId = blueprintId.ifBlank { "default" },
            title = title,
            description = description,
            agentId = agentId,
            priority = priority.name,
            status = status.name,
            dependsOn = json.encodeToString(dependsOn),
            timeoutSeconds = timeoutSeconds,
            retryCount = retryCount,
            maxRetries = maxRetries,
            result = result?.let { json.encodeToString(it) },
            error = error,
            createdAt = createdAt,
            updatedAt = updatedAt
        )
    }
    
    private fun TaskEntity.toDomain(): Task {
        return Task(
            id = id,
            blueprintId = blueprintId,
            title = title,
            description = description,
            agentId = agentId,
            priority = try { Priority.valueOf(priority) } catch (e: Exception) { Priority.MEDIUM },
            status = try { TaskStatus.valueOf(status) } catch (e: Exception) { TaskStatus.QUEUED },
            dependsOn = try { json.decodeFromString(dependsOn) } catch (e: Exception) { emptyList() },
            timeoutSeconds = timeoutSeconds,
            retryCount = retryCount,
            maxRetries = maxRetries,
            result = result?.let { try { json.decodeFromString(it) } catch (e: Exception) { null } },
            error = error,
            createdAt = createdAt,
            updatedAt = updatedAt
        )
    }
}

File: pc-bridge/src/database/repositories/TaskRepository.ts
// SPDX-License-Identifier: MIT
/**
 * Task Repository – CRUD für Orchestrierungs-Tasks.
 *
 * ## Dependency-Graph:
 * depends_on wird als JSON-Array gespeichert.
 * Der Orchestrator prüft vor dem Dispatch ob alle Dependencies "done" sind.
 */

import type BetterSqlite3 from 'better-sqlite3';

export type TaskStatus = 'queued' | 'dispatching' | 'working' | 'reviewing' | 'done' | 'failed' | 'blocked' | 'cancelled';
export type Priority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface Task {
  id: string;
  sessionId: string | null;
  blueprintId: string | null;
  title: string;
  description: string;
  agentId: string;
  priority: Priority;
  status: TaskStatus;
  dependsOn: string[];
  timeoutSeconds: number;
  retryCount: number;
  maxRetries: number;
  resultType: string | null;
  resultContent: string | null;
  error: string | null;
  createdAt: number;
  updatedAt: number;
  metadata: Record<string, unknown>;
}

export class TaskRepository {
  constructor(private db: BetterSqlite3.Database) {}

  create(task: Omit<Task, 'createdAt' | 'updatedAt'>): Task {
    const now = Math.floor(Date.now() / 1000);
    this.db.prepare(`
      INSERT INTO tasks (id, session_id, blueprint_id, title, description, agent_id, priority, status, depends_on, timeout_seconds, retry_count, max_retries, result_type, result_content, error, created_at, updated_at, metadata)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(
      task.id,
      task.sessionId,
      task.blueprintId,
      task.title,
      task.description,
      task.agentId,
      task.priority,
      task.status,
      JSON.stringify(task.dependsOn),
      task.timeoutSeconds,
      task.retryCount,
      task.maxRetries,
      task.resultType,
      task.resultContent,
      task.error,
      now,
      now,
      JSON.stringify(task.metadata)
    );
    return { ...task, createdAt: now, updatedAt: now };
  }

  getById(id: string): Task | null {
    const row = this.db.prepare('SELECT * FROM tasks WHERE id = ?').get(id) as any;
    if (!row) return null;
    return this.mapRow(row);
  }

  listByStatus(status: TaskStatus): Task[] {
    const rows = this.db.prepare('SELECT * FROM tasks WHERE status = ? ORDER BY created_at ASC').all(status) as any[];
    return rows.map((r) => this.mapRow(r));
  }

  listBySession(sessionId: string): Task[] {
    const rows = this.db.prepare('SELECT * FROM tasks WHERE session_id = ? ORDER BY created_at ASC').all(sessionId) as any[];
    return rows.map((r) => this.mapRow(r));
  }

  listByBlueprint(blueprintId: string): Task[] {
    const rows = this.db.prepare('SELECT * FROM tasks WHERE blueprint_id = ? ORDER BY created_at ASC').all(blueprintId) as any[];
    return rows.map((r) => this.mapRow(r));
  }

  updateStatus(id: string, status: TaskStatus): void {
    const now = Math.floor(Date.now() / 1000);
    this.db.prepare('UPDATE tasks SET status = ?, updated_at = ? WHERE id = ?').run(status, now, id);
  }

  setResult(id: string, resultType: string, resultContent: string): void {
    const now = Math.floor(Date.now() / 1000);
    this.db.prepare('UPDATE tasks SET status = ?, result_type = ?, result_content = ?, updated_at = ? WHERE id = ?')
      .run('done', resultType, resultContent, now, id);
  }

  setError(id: string, error: string): void {
    const now = Math.floor(Date.now() / 1000);
    this.db.prepare('UPDATE tasks SET status = ?, error = ?, updated_at = ? WHERE id = ?')
      .run('failed', error, now, id);
  }

  incrementRetry(id: string): void {
    const now = Math.floor(Date.now() / 1000);
    this.db.prepare('UPDATE tasks SET retry_count = retry_count + 1, status = ?, updated_at = ? WHERE id = ?')
      .run('queued', now, id);
  }

  /**
   * Prüft ob alle Dependencies eines Tasks "done" sind.
   */
  areDependenciesMet(task: Task): boolean {
    if (task.dependsOn.length === 0) return true;
    const placeholders = task.dependsOn.map(() => '?').join(',');
    const rows = this.db.prepare(
      `SELECT id, status FROM tasks WHERE id IN (${placeholders})`
    ).all(...task.dependsOn) as any[];
    return rows.length === task.dependsOn.length && rows.every((r) => r.status === 'done');
  }

  delete(id: string): void {
    this.db.prepare('DELETE FROM tasks WHERE id = ?').run(id);
  }

  private mapRow(row: any): Task {
    return {
      id: row.id,
      sessionId: row.session_id,
      blueprintId: row.blueprint_id,
      title: row.title,
      description: row.description,
      agentId: row.agent_id,
      priority: row.priority,
      status: row.status,
      dependsOn: JSON.parse(row.depends_on || '[]'),
      timeoutSeconds: row.timeout_seconds,
      retryCount: row.retry_count,
      maxRetries: row.max_retries,
      resultType: row.result_type,
      resultContent: row.result_content,
      error: row.error,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
      metadata: JSON.parse(row.metadata || '{}'),
    };
  }
}
